### PR TITLE
Update `prost`, `prost-derive`, `pbjson`, `tonic` ecosystem

### PR DESCRIPTION
**Pull request by alamb**
**NOTE**: If we merge this PR before arrow-flight is released with an upgraded version of prost, I think DataFusion based systems might not be able to upgrade DataFusion until that new flight version is available

I didn't look carefully, but if I updated prost/tonic in datafusion-examples (needed for the flight server example) I got a bunch of compiler errors

## Which issue does this PR close?

Closes https://github.com/apache/datafusion/pull/11352
closes https://github.com/apache/datafusion/pull/11353
and several other dependency PR updates

## Rationale for this change

Keep up to date with dependencies

### CODE REVIEW

**Comment by alamb** [comment]
🤔  looks like maybe we need to update arrow-flight's dependencies too 🤔

File: datafusion/substrait/Cargo.toml [hunk 1]
```diff
@@ -38,8 +38,8 @@ chrono = { workspace = true }
 datafusion = { workspace = true, default-features = true }
 itertools = { workspace = true }
 object_store = { workspace = true }
-pbjson-types = "0.6"
-prost = "0.12"
+pbjson-types = "0.7"
+prost = "0.13"
 substrait = { version = "0.36.0", features = ["serde"] }
```
Comment (tisonkun) [inline datafusion/substrait/Cargo.toml:43]:
```suggestion
substrait = { version = "0.37.0", features = ["serde"] }
```

Comment (tisonkun) [reply datafusion/substrait/Cargo.toml:43]: I tested locally and this can work with `cargo build --workspace`.

**Comment by alamb** [comment]
Thank you @tisonkun 

It appears to me that compilation is failing now as arrow-flight is using an older version of prost/tonic, but since datafusion itself doesn't depend on any of those, I reverted the upgrade

File: datafusion/proto/src/generated/datafusion_proto_common.rs [hunk 2]
```diff
@@ -722,19 +722,19 @@ pub mod column_options {
...
     #[allow(clippy::derive_partial_eq_without_eq)]
-    #[derive(Clone, PartialEq, ::prost::Oneof)]
+    #[derive(Clone, Copy, PartialEq, ::prost::Oneof)]
     pub enum BloomFilterFppOpt {
         #[prost(double, tag = "6")]
         BloomFilterFpp(f64),
     }
     #[allow(clippy::derive_partial_eq_without_eq)]
-    #[derive(Clone, PartialEq, ::prost::Oneof)]
+    #[derive(Clone, Copy, PartialEq, ::prost::Oneof)]
```
Comment (comphead) [inline datafusion/proto/src/generated/datafusion_proto_common.rs:731]: just wondering why the Copy was added?

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb

**Comment by alamb** [comment]
I believe if we merge this PR, people won't be able to upgrade DataFusion with their systems that use arrow-flight  unil this is also released: https://github.com/apache/arrow-rs/pull/6041

File: datafusion/proto/src/generated/datafusion_proto_common.rs (hunk 2 above)
Comment (alamb) [reply datafusion/proto/src/generated/datafusion_proto_common.rs:731]: This is  generated code. so I think it comes from the changes in prost. I think it comes from https://github.com/tokio-rs/prost/pull/950 specifically

**Comment by alamb** [comment]
To be clear, this is blocked on the release of arrow-flight `53.0.0`, eta in about 3 weeks

**Comment by alamb** [comment]
Superceded by https://github.com/apache/datafusion/pull/12032

**Pull request closed without merging**